pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -229,26 +235,66 @@ func (m *ManagerImpl) Start(activePods ActivePodsFunc, sourcesReady config.Sourc
 	return nil
 }
 
-// GetWatcherCallback returns callback function to be registered with plugin watcher
-func (m *ManagerImpl) GetWatcherCallback() watcher.RegisterCallbackFn {
+// GetWatcherHandler returns the plugin handler
+func (m *ManagerImpl) GetWatcherHandler() watcher.PluginHandler {
 	if f, err := os.Create(m.socketdir + "DEPRECATION"); err != nil {
 		glog.Errorf("Failed to create deprecation file at %s", m.socketdir)
 	} else {
 		f.Close()
 		glog.V(4).Infof("created deprecation file %s", f.Name())
 	}
 
-	return func(name string, endpoint string, versions []string, sockPath string) (chan bool, error) {
-		if !m.isVersionCompatibleWithPlugin(versions) {
-			return nil, fmt.Errorf("manager version, %s, is not among plugin supported versions %v", pluginapi.Version, versions)
-		}
+	return watcher.PluginHandler(m)
+}
 
-		if !v1helper.IsExtendedResourceName(v1.ResourceName(name)) {
-			return nil, fmt.Errorf("invalid name of device plugin socket: %v", fmt.Sprintf(errInvalidResourceName, name))
-		}
+// ValidatePlugin validates a plugin if the version is correct and the name has the format of an extended resource
+func (m *ManagerImpl) ValidatePlugin(pluginName string, endpoint string, versions []string) error {
+	if !m.isVersionCompatibleWithPlugin(versions) {
+		return fmt.Errorf("manager version, %s, is not among plugin supported versions %v", pluginapi.Version, versions)
+	}
+
+	if !v1helper.IsExtendedResourceName(v1.ResourceName(pluginName)) {
+		return fmt.Errorf("invalid name of device plugin socket: %v", fmt.Sprintf(errInvalidResourceName, pluginName))
+	}
+
+	return nil
+}
+
+// InitPlugin starts the endpoint
+// TODO: Start the endpoint and wait for the First ListAndWatch call
+//       before swiching to the Register state
+func (m *ManagerImpl) InitPlugin(pluginName string, endpoint string) error {
+	e, err := newEndpointImpl(endpoint, pluginName, m.callback)
+	if err != nil {
+		return fmt.Errorf("Failed to dial device plugin with socketPath %s: %v", endpoint, err)
+	}
 
-		return m.addEndpointProbeMode(name, sockPath)
+	options, err := e.client.GetDevicePluginOptions(context.Background(), &pluginapi.Empty{})
+	if err != nil {
+		return fmt.Errorf("Failed to get device plugin options: %v", err)
 	}
+
+	// We can do this because the pluginwatcher module guarantees that no
+	// other plugins with the same name will register before the `Register`
+	// call returns for that pluginName
+	m.endpointsRegistering[pluginName] = endpointInfo{e: e, opts: options}
+
+	return nil
+}
+
+// RegisterPlugin registers the plugin and starts running it
+func (m *ManagerImpl) RegisterPlugin(pluginName, endpoint string) {
+	eI := m.endpointsRegistering[pluginName]
+	delete(m.endpointsRegistering, pluginName)
+
+	m.registerEndpoint(pluginName, eI.opts, eI.e)
+	go m.runEndpoint(pluginName, eI.e)
```

Comment:
> You only stressed to enforce this on my initial PR and to address that i introduced chanForAckOfNotification

I agree, my stance has changed a bit since then. In both case I think what's important is that the plugin gets notified at a defined state in the lifecycle, not at an undefined time.

Here the lifecycle is defined as follows for the Plugin Consumer:
```
Validate +----------> Init +---------> Register
```

After Init finishes the plugin will get that notification in a synchronous fashion. Once that notification has been acknowledged by the plugin, the plugin consumer can then start to consume the plugin (through the register call).

Here the init phase allows the device manager to setup the device plugin before starting to use it (e.g: Advertise the number of devices). This would translate in my model to calling `GetDevicePluginOptions` but also start the stream to `ListAndWatch` to get the devices.

This would guarantee the device manager that plugins that go through the `Register` call have already sent their devices.

Reply:
not sure what you mean by "undefined time". If plugin gets notified at "Registered" state, i think that should be called as defined.
```
validated ---> initialized --> registered---> notified --> deregistered
```
instead of 

```
validated ---> initialized --> notified ---> registered ---> deregistered
```

> Here the init phase allows the device manager to setup the device plugin before starting to use it (e.g: Advertise the number of devices). This would translate in my model to calling GetDevicePluginOptions but also start the stream to ListAndWatch to get the devices.

This does not look clean to me. IMO, `GetDevicePluginOptions` or any other consumer specific communication should start after handshake between watcher and plugin is complete and communication is handed over to consumer.

Reply:
> not sure what you mean by "undefined time". If plugin gets notified at "Registered" state, i think that should be called as defined.
> validated ---> initialized --> registered---> notified --> deregistered

I think we are both saying the same thing for the same path.

> This does not look clean to me. IMO, GetDevicePluginOptions or any other consumer specific communication should start after handshake between watcher and plugin is complete and communication is handed over to consumer.

In a 0 downtime setting, how would a plugin know when it is being used? i.e:
If I am plugin v1.10 for resource `foo.io/bar`, how do I know when kubelet stopped using plugin v1.9?

Reply:
once you(1.10 plugin) have received a success notification, you should be ready to serve apis like  GetDevicePluginOptions, ListAndWatch etc.

By 0 downtime did you mean the period after success notification to the plugin and before receiving devices on ListAndWatch? If yes, i am not sure if it is a expected upgrade strategy to upgrade the node while keeping it online and allowing pods to get scheduled during upgrade.

Reply:
> By 0 downtime did you mean the period after success notification to the plugin and before receiving devices on ListAndWatch?
> If yes, i am not sure if it is a expected upgrade strategy to upgrade the node while keeping it online and allowing pods to get scheduled during upgrade.

Upgrading a Device Plugin doesn't happen only when the Device Plugin API changes or when the node upgrades.

Plugins themselves might have have a versioning scheme that though it will always be dependent on the Device Plugin API (in the device plugin case) can also publish new releases.
I believe it is reasonable to expect that the plugin watcher provides the mechanisms to handle 0 downtime (they are in this PR).
It is also reasonable to expect the Device Plugins to handle 0 downtime upgrade (it'll be the case of GPUs) if they want / need to.

How would you implement that requirement differently?

Reply:
it seems only functionality change is 0 downtime in this PR and rest is refactoring, right?

Reply:
- Deregister Handling:
  - I just noticed I could at least stop the endpoint on the device plugin we can discuss of removing the plugin in a followup PR
  - On the CSI side this is still in discussion and will happen in a followup PR
- ReRegistration Handling  (plugin watcher level)
- Use of an interface as the kubelet facing API
- Some refactor as you pointed out